### PR TITLE
CORTX-32800:validate min values of rgw resource limits provided in Gconf

### PR DESCRIPTION
**Pull request by sachitanands**
Validate min values of CPU/Memory of rgw resource limits.
If those values are less than our expected lower limits then mark that deployment as failure.
In rgw, we have to set some default values for these minimum value as limit value.
User should add same or more than these values in solution.yaml file.
 

Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: src/rgw/const.py [hunk 1]
```diff
@@ -141,6 +141,24 @@
 UTF_ENCODING = 'utf-8'
 MOTR_MY_FID = f'{SVC_SECTION}>motr my fid'
 
+# RGW resource limit keys
+SVC_CPU_MIN_VAL_LIMIT = '250m'
```
Comment (sachitanands) [inline src/rgw/const.py:145]: these limit values are currently copied from existing solution.yaml file.
Saurabh is exploring on exact values in current sprint and we will update these values in Sprint 75 with more specific values.

File: src/rgw/setup/rgw.py [hunk 2]
```diff
@@ -422,10 +426,10 @@ def _create_rgw_user(conf: MappedConf):
         except CipherInvalidToken as e:
             raise SetupError(errno.EINVAL, f'auth_secret decryption failed. {e}')
         rgw_config = Rgw._get_rgw_config_path(conf)
-        create_usr_cmd = f'radosgw-admin user create --uid={user_name} --access-key \
-            {access_key} --secret {password} --display-name="{user_name}" \
-            --caps="users=*;metadata=*;usage=*;zone=*;info=*;user=*;roles=*;user-policy=*;buckets=*" \
-            -c {rgw_config} -n client.radosgw-admin --no-mon-config'
+        create_usr_cmd = (f'radosgw-admin user create --uid={user_name} --access-key '
```
Comment (sachitanands) [inline src/rgw/setup/rgw.py:429]: handled some formatting changes found during debugging

File: src/rgw/const.py (hunk 1 above)
Comment (sjain09) [reply src/rgw/const.py:145]: @sachitanands , are supporting cpu limit values without any unit specified by user? for example cpu : 1 is also a valid value.

File: src/rgw/const.py [hunk 3]
```diff
@@ -141,6 +141,24 @@
...
+SVC_MEM_MIN_VAL_LIMIT = '128Mi'
+SVC_LIMIT_KEY = f'cortx>{COMPONENT_NAME}>limits'
+SVC_LIMIT_NUM_SERVICES= f'{SVC_LIMIT_KEY}>num_services'
+SVC_LIMIT_NAME = f'{SVC_LIMIT_KEY}>services[%s]>name'
+SVC_LIMIT_CPU_MIN_KEY = f'{SVC_LIMIT_KEY}>services[%s]>cpu>min'
+SVC_LIMIT_MEM_MIN_KEY = f'{SVC_LIMIT_KEY}>services[%s]>memory>min'
+SVC_RESOURCE_LIMIT_MEM_VAL_SUFFIXES = ['K', 'Ki', 'Kib', 'M', 'Mi', 'Mib', 'G', 'Gi', 'Gib']
+SVC_RESOURCE_LIMIT_MEM_VAL_SIZE_MAP = {
+        "K": 1024, "M": 1024*1024, "G": 1024*1024*1024,
```
Comment (cdeshmukh) [inline src/rgw/const.py:154]: For Ki, Kib units (& alike) why cant we match values against K, we dont need keep duplicated values for possible variants of same unit

File: src/rgw/setup/rgw.py [hunk 4]
```diff
@@ -852,6 +856,82 @@ def _verify_backend_store_value(conf: MappedConf):
                 f'Supported rgw backend store are {const.SUPPORTED_BACKEND_STORES},'
                 f' currently configured one is {backend_store}')
 
+    @staticmethod
+    def _validate_resource_limit_values(conf: MappedConf):
+        """Validate minimum values of rgw resource limits in Gconf."""
+        Log.info(f'Validating minimum values of resource limits for {const.COMPONENT_NAME}.')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:862]: Validating configured values .... against minimum required values

File: src/rgw/setup/rgw.py [hunk 5]
```diff
@@ -852,6 +856,82 @@ def _verify_backend_store_value(conf: MappedConf):
...
+        num_services = int(Rgw._get_cortx_conf(conf, const.SVC_LIMIT_NUM_SERVICES))
+        if num_services == 0:
+            raise SetupError(errno.EINVAL,
+                f'Invalid/Missing values found in gconf for key :{const.SVC_LIMIT_NUM_SERVICES}')
+        input_cpu_min_val = ''
+        input_mem_min_val = ''
+        for value_index in range(0, num_services):
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:869]: I really dont like this  looping over, to find rgw service resource limits. Doesnt conf store give a way to make it available as dictionary, where on feeding 'rgw' key, we get the values ?
latest solution.yaml https://github.com/Seagate/cortx-k8s/blob/v0.9.0/k8_cortx_cloud/solution.example.yaml

File: src/rgw/setup/rgw.py [hunk 6]
```diff
@@ -852,6 +856,82 @@ def _verify_backend_store_value(conf: MappedConf):
...
+            converted_input_val = Rgw._convert_resource_limit_value(input_val, limit_type)
+
+        if expected_val.isnumeric():
+            converted_expected_val = int(expected_val)
+        else:
+            converted_expected_val = Rgw._convert_resource_limit_value(expected_val, limit_type)
+
+        if converted_input_val < converted_expected_val :
+            raise SetupError(errno.EINVAL,
+                f'Provided value {input_val} for rgw resource limit is less than expected value {expected_val}')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:898]: For what limit_type, cpu/mem that should also be logged

File: src/rgw/setup/rgw.py [hunk 7]
```diff
@@ -852,6 +856,82 @@ def _verify_backend_store_value(conf: MappedConf):
...
+
+    @staticmethod
+    def _convert_resource_limit_value(resource_limit_val: str, limit_type: str):
+        """"Convert give resource limit value to common units based on limit type"""
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:902]: typo give  -> given

File: src/rgw/setup/rgw.py [hunk 8]
```diff
@@ -852,6 +856,82 @@ def _verify_backend_store_value(conf: MappedConf):
...
+
+            # Ex: If mem resource_limit_val is 128MiB then map_val = 1024*1024*1024 or
+            # If cpu resource_limit_val is 200m then map_val = 1
+            if limit_type == 'mem' :
+                map_val = const.SVC_RESOURCE_LIMIT_MEM_VAL_SIZE_MAP[suffix]
+            else :
+                map_val = const.SVC_RESOURCE_LIMIT_CPU_VAL_SIZE_MAP[suffix]
+
+            # Calcuate final limit value.
+            ret = int(num_resource_limit_val) * int(map_val)
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:927]: map_val is already int, but conf store should give that way, otherwise this int conversion doesnt hurt

Comment (sachitanands) [reply src/rgw/setup/rgw.py:927]: map_val is stored in our code cosnt.py so its safe to remove conversion.. 
will update it.

File: src/rgw/setup/rgw.py (hunk 5 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:869]: As of now, consul based confstore does not provide any way to return any list.
hence we need to fetch num_services key and then iterate over each keys using count(value_index) variable and then fetch actual value.

Regarding latest gconf keys in solution.yaml..
below is updated copy which is being used in deployment 
( https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/651132980/CORTX+GConf+Config+Store+Keys#Example%3A-CORTX-GConf-Part-A-(%2Fetc%2Fcortx%2Fcluster.conf) Line : 176 -185

Today we have only one service (service : rgw) for which this limits are defined but there can be multiple services for which limit will be added. hence to differentiate between rgw and other service, I have added a check on service name parameter while iterating over idex.

File: src/rgw/setup/rgw.py (hunk 8 above)
Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:927]: oh const it is, i read conf

File: src/rgw/const.py (hunk 1 above)
Comment (sachitanands) [reply src/rgw/const.py:145]: yes. added a logic to handle this numeric value. (e.g. 1 cpu = 1000m) any fractional value will also be handled in similar way.

**Comment by sachitanands** [comment]
Did the deployment with latest changes,

Deployment job : https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Generic/job/rgw-integration/74/ 
 Docker build  :

`cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-74-rgw-integration-pr
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-74-rgw-integration-pr
cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-74-rgw-integration-pr
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-74-rgw-integration-pr
`

Cluster deployment job : [1Node-deployment.PNG](https://jts.seagate.com/secure/attachment/528594/528594_1Node-deployment.PNG) 
     https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/9657/  

Ran basic sanity work by changing limit values with both valid n invalid cases
valid values : [valid_success_case.PNG](https://jts.seagate.com/secure/attachment/528598/528598_valid_success_case.PNG) 
Invalid cpu values : [invalid_cpu_val_failure_msg.PNG](https://jts.seagate.com/secure/attachment/528595/528595_invalid_cpu_val_failure_msg.PNG) 
Invalid mem values : [invalid_value_failure.PNG](https://jts.seagate.com/secure/attachment/528596/528596_invalid_value_failure.PNG)

File: src/rgw/setup/rgw.py (hunk 5 above)
Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:869]: alright

**Review by cdeshmukh — APPROVED** [review]

File: src/rgw/setup/rgw.py [hunk 9]
```diff
@@ -852,6 +856,92 @@ def _verify_backend_store_value(conf: MappedConf):
...
+            temp = list(filter(resource_limit_val.endswith, const.SVC_RESOURCE_LIMIT_CPU_VAL_SUFFIXES))
+        else:
+            raise SetupError(errno.EINVAL, f'Invalid resource limit type {limit_type} speicified for rgw.')
+        if len(temp) > 0:
+            # Ex: If mem resource_limit_val is 128MiB then num_resource_limit_val=128 or
+            # If cpu resource_limit_val is 200m then num_resource_limit_val=200
+            num_resource_limit_val = re.sub(r'[^0-9]', '', resource_limit_val)
+            val_length = len(num_resource_limit_val)
+            # If mem resource_limit_val is 128MiB then resource_unit_key is M.
+            resource_unit_key = resource_limit_val[val_length:val_length+1]
```
[the remark this answers is not in the document]
Comment (sachitanands) [reply src/rgw/setup/rgw.py:927]: this will be done in 916 , 918 Line. 
if input string is not valid then it will raise SetupError at 940 line

**Review by shriya-deshmukh — APPROVED** [review]

**Review by priyanka25081999 — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**